src/main/java/com/azure/cosmos/kafka/connect/sink/id/strategy/TemplateStrategy.java

```diff
@@ -0,0 +1,66 @@
+package com.azure.cosmos.kafka.connect.sink.id.strategy;
+
+import com.azure.cosmos.implementation.guava25.collect.ImmutableMap;
+import org.apache.kafka.connect.data.Values;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+import java.util.Map;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class TemplateStrategy extends AbstractIdStrategy {
+    private static final String KEY = "key";
+    private static final String VALUE = "value";
+    private static final String TOPIC = "topic";
+    private static final String PARTITION = "partition";
+    private static final String OFFSET = "offset";
+
+    private static final String PATTERN_TEMPLATE = "\\$\\{(%s)\\}";
+
+    private TemplateStrategyConfig config;
+
+    private static final Map<String, Function<SinkRecord, String>> METHODS_BY_VARIABLE;
+
+    static {
+        ImmutableMap.Builder<String, Function<SinkRecord, String>> builder = ImmutableMap.builder();
+        builder.put(KEY, (r) -> Values.convertToString(r.keySchema(), r.key()));
+        builder.put(VALUE, (r) -> Values.convertToString(r.valueSchema(), r.value()));
+        builder.put(TOPIC, SinkRecord::topic);
+        builder.put(PARTITION, (r) -> r.kafkaPartition().toString());
+        builder.put(OFFSET, (r) -> Long.toString(r.kafkaOffset()));
+        METHODS_BY_VARIABLE = builder.build();
+    }
+
+    @Override
+    public String generateId(SinkRecord record) {
+        String template = config.template();
+        return resolveAll(template, record);
+    }
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        config = new TemplateStrategyConfig(configs);
+
+        super.configure(configs);
+    }
+
+    private String resolveAll(String template, SinkRecord record) {
+        String pattern = String.format(PATTERN_TEMPLATE,
+                METHODS_BY_VARIABLE.keySet().stream().collect(Collectors.joining("|")));
+        int lastIndex = 0;
+        StringBuilder output = new StringBuilder();
+        Matcher matcher = Pattern.compile(pattern).matcher(template);
```

Comment:
how often is this method invoked? for every item to be ingested? pattern matching generally is very expensive. Is there any other way we can do this? cache regex, etc?

Reply:
This would be invoked on each sink record. We can do our due diligence and create a static `Pattern` that we reuse for each record, but since all variables need to be "bound" to fields on the record, we kind of have to do the regex matching in the hot path. We also do this, albeit at a simpler level in other connectors like [JDBC](https://github.com/confluentinc/kafka-connect-jdbc/blob/83b62f2873d8627a781d5f100c35d1a136671e15/src/main/java/io/confluent/connect/jdbc/sink/JdbcDbWriter.java#L100). The `replace` behind the scenes just does a `Pattern.compile().matcher().replaceAll()` . We could go the same route and use `replace` instead, but I don't know if there would be much improvement there.

I'm inclined to cache the compiled Pattern so we don't recreate it with each record, but keep the usage of matcher the same. My understanding is the compile should be the expensive part as it creates the state machine. Matching after that should not be that expensive? WDYT @moderakh ?